### PR TITLE
[calyx-py] Minor builder extensions

### PR DESCRIPTION
**Pull request by EclecticGriffin**
Small batch of changes to improve the builder.

- add a `sub_component` constructor for `ComponentBuilder` to create a cell instantiating a Calyx sub-comp
- add a `get_component` method to `Builder` to lookup `ComponentBuilder`'s by name
- extend the `invoke` constructor function to take `ref_*` arguments
- misc docstrings and type labels

Of these it's mostly the `invoke` thing that was needed by @anshumanmohan

### CODE REVIEW

File: calyx-py/calyx/builder.py [hunk 1]
```diff
@@ -131,7 +142,25 @@ def cell(
         self.index[name] = builder
         return builder
 
-    def reg(self, name: str, size: int):
+    def sub_component(
+        self,
+        cell_name: str,
+        sub_comp_name: str | ComponentBuilder,
```
Comment (sampsyo) [inline calyx-py/calyx/builder.py:148]: I know this is silly, but maybe this should just be called `comp_name` (and the docstring should just say "component")? I don't think we've used "sub-component" elsewhere, and I'm not sure what it means… we don't exactly have a hierarchy of components in Calyx (although we do have a hierarchy of cells).

File: calyx-py/calyx/builder.py [hunk 2]
```diff
@@ -221,7 +250,7 @@ def if_(port: ExprBuilder, cond: Optional[GroupBuilder], body) -> ast.If:
 def invoke(cell: CellBuilder, **kwargs) -> ast.Invoke:
     """Build an `invoke` control statement.
 
-    The keyword arguments should have the form `in_*` and `out_*`, where
+    The keyword arguments should have the form `in_*`, `out_*`, or `ref_*`, where
     `*` is the name of an input or output port on the invoked cell.
```
Comment (sampsyo) [inline calyx-py/calyx/builder.py:254]: input port, output port, or ref cell

**Review by sampsyo — COMMENTED** [review]
Neato!! Just a couple of small cosmetic suggestions!

File: calyx-py/calyx/builder.py (hunk 1 above)
Comment (EclecticGriffin) [reply calyx-py/calyx/builder.py:148]: How do you feel about `comp_instance`?

Comment (sampsyo) [reply calyx-py/calyx/builder.py:148]: Good call!

**Pull request merged**